ClassCastException when trying to get Element from Node

### Describe the bug

Hi,
I'm using aws-cdk for java. After upgrading the artifact software.amazon.jsii:jsii-runtime from version 1.85.0 to 1.86.0 a lot of my stacks don't compile anymore. Here an example stack
```
public class RolesAnywhereStack extends Stack {
    public RolesAnywhereStack(App parent) {
        super(parent, "RolesAnywhereStack");
        final Role bastionRole = Role.Builder.create(this, "RolesAnywhereBastionRole")
                .assumedBy(new ServicePrincipal("rolesanywhere.amazonaws.com"))
                .description("Role that is used when connecting via RolesAnywhere")
                .build();
        CfnRole cfnRole = (CfnRole) bastionRole.getNode().getDefaultChild();
    }
}
```
when running `cdk synthesize RolesAnywhereStack` a ClassCastException occurs:
```
Caused by: java.lang.ClassCastException: class software.constructs.Construct cannot be cast to class software.amazon.awscdk.services.iam.CfnRole (software.constructs.Construct and software.amazon.awscdk.services.iam.CfnRole are in unnamed module of loader java.net.URLClassLoader @7486e0ca)
```

with jsii-runtime 1.85.0 everything works fine.

### Expected Behavior

No ClassCastException

### Current Behavior

ClassCastException

### Reproduction Steps

try to synthesize this stack
```
public class RolesAnywhereStack extends Stack {
    public RolesAnywhereStack(App parent) {
        super(parent, "RolesAnywhereStack");
        final Role bastionRole = Role.Builder.create(this, "RolesAnywhereBastionRole")
                .assumedBy(new ServicePrincipal("rolesanywhere.amazonaws.com"))
                .description("Role that is used when connecting via RolesAnywhere")
                .build();
        CfnRole cfnRole = (CfnRole) bastionRole.getNode().getDefaultChild();
    }
}
```


### Possible Solution

_No response_

### Additional Information/Context

_No response_

### SDK version used

software.amazon.jsii:jsii-runtime:1.85.0

### Environment details (OS name and version, etc.)

Debian openjdk version "17.0.2" 2022-01-18 OpenJDK Runtime Environment (build 17.0.2+8-86) OpenJDK 64-Bit Server VM (build 17.0.2+8-86, mixed mode, sharing)

--- a/packages/@jsii/kernel/src/objects.test.ts
+++ b/packages/@jsii/kernel/src/objects.test.ts
@@ -1,4 +1,4 @@
-import { ObjectTable } from './objects';
+import { ObjectTable, jsiiTypeFqn } from './objects';
 
 const mockResolve = jest.fn();
 
@@ -23,4 +23,21 @@ test('registered objects have clean JSON.Stringify', () => {
   subject.registerObject(obj, 'Object');
 
   expect(JSON.stringify(obj)).toEqual(expected);
+});
+
+test('FQN resolution is not broken by inheritance relationships', () => {
+  const rtti = Symbol.for('jsii.rtti');
+
+  class Parent {
+    public static readonly [rtti] = { fqn: 'phony.Parent' };
+  }
+  class Child extends Parent {
+    public static readonly [rtti] = { fqn: 'phony.Child' };
+  }
+
+  const parent = new Parent();
+  expect(jsiiTypeFqn(parent, () => true)).toBe(Parent[rtti].fqn);
+
+  const child = new Child();
+  expect(jsiiTypeFqn(child, () => true)).toBe(Child[rtti].fqn);
 });

--- a/packages/@jsii/kernel/src/objects.ts
+++ b/packages/@jsii/kernel/src/objects.ts
@@ -16,30 +16,14 @@ const OBJID_SYMBOL = Symbol.for('$__jsii__objid__$');
 const IFACES_SYMBOL = Symbol.for('$__jsii__interfaces__$');
 
 /**
- * Symbol we use to tag constructors that are exported from a jsii module.
- */
-const JSII_TYPE_FQN_SYMBOL = Symbol('$__jsii__fqn__$');
-
-/**
  * Symbol under which jsii runtime type information is stored.
  */
 const JSII_RTTI_SYMBOL = Symbol.for('jsii.rtti');
 
 /**
- * Exported constructors processed by the jsii compiler have runtime type
- * information woven into them under the `[JSII_RTTI]` symbol property.
+ * Cache for resolved associations between constructors and FQNs.
  */
-interface MaybeManagedConstructor {
-  readonly [JSII_RTTI_SYMBOL]?: {
-    /** The fully qualified name of the jsii type. */
-    readonly fqn: spec.FQN;
-    /** The version of the assembly that declared this type. */
-    readonly version: string;
-  };
-
-  /** The resolved JSII type FQN for this type. This is set only after resolution */
-  readonly [JSII_TYPE_FQN_SYMBOL]?: spec.FQN;
-}
+const RESOLVED_TYPE_FQN = new WeakMap<object, spec.FQN>();
 
 /**
  * Get the JSII fqn for an object (if available)
@@ -55,11 +39,11 @@ export function jsiiTypeFqn(
   obj: any,
   isVisibleType: (fqn: spec.FQN) => boolean,
 ): spec.FQN | undefined {
-  const ctor = obj.constructor as MaybeManagedConstructor;
+  const ctor = obj.constructor;
 
   // We've already resolved for this type, return the cached value.
-  if (ctor[JSII_TYPE_FQN_SYMBOL] != null) {
-    return ctor[JSII_TYPE_FQN_SYMBOL];
+  if (RESOLVED_TYPE_FQN.has(ctor)) {
+    return RESOLVED_TYPE_FQN.get(ctor)!;
   }
 
   let curr = ctor;
@@ -138,22 +122,19 @@ function tagObject(obj: unknown, objid: string, interfaces?: string[]) {
  * Set the JSII FQN for classes produced by a given constructor
  */
 export function tagJsiiConstructor(constructor: any, fqn: spec.FQN) {
-  if (Object.prototype.hasOwnProperty.call(constructor, JSII_TYPE_FQN_SYMBOL)) {
+  const existing = RESOLVED_TYPE_FQN.get(constructor);
+
+  if (existing != null) {
     return assert.strictEqual(
-      constructor[JSII_TYPE_FQN_SYMBOL],
+      existing,
       fqn,
-      `Unable to register ${constructor.name} as ${fqn}: it is already registerd with FQN ${constructor[JSII_TYPE_FQN_SYMBOL]}`,
+      `Unable to register ${constructor.name} as ${fqn}: it is already registerd with FQN ${existing}`,
     );
   }
 
   // Mark this constructor as exported from a jsii module, so we know we
   // should be considering it's FQN as a valid exported type.
-  Object.defineProperty(constructor, JSII_TYPE_FQN_SYMBOL, {
-    configurable: false,
-    enumerable: false,
-    writable: false,
-    value: fqn,
-  });
+  RESOLVED_TYPE_FQN.set(constructor, fqn);
 }
 
 /**

